internal/service/service.go

```diff
@@ -0,0 +1,159 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package service
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/observiq/observiq-otel-collector/collector"
+	"go.uber.org/zap"
+)
+
+const (
+	startTimeout = 10 * time.Second
+	stopTimeout  = 10 * time.Second
+)
+
+// RunnableService may be run as a service.
+type RunnableService interface {
+	// Start asynchronously starts the underlying service. The service may not necessarily be "ready"
+	// once this returns, but could be asynchronously starting up.
+	Start(ctx context.Context) error
+	// Stop synchronously shuts down the service. After this function returns, the underlying service should be completely stopped.
+	Stop(ctx context.Context) error
+	// Error returns an error channel that should emit an error when the service must unexpectedly quit.
+	Error() <-chan error
+}
+
+// StandaloneCollectorService is a RunnableService that runs the collector in standalone mode.
+type StandaloneCollectorService struct {
+	col      collector.Collector
+	doneChan chan struct{}
+	errChan  chan error
+	wg       *sync.WaitGroup
+}
+
+// NewStandaloneCollectorService creates a new StandaloneCollectorService
+func NewStandaloneCollectorService(c collector.Collector) StandaloneCollectorService {
+	return StandaloneCollectorService{
+		col:      c,
+		doneChan: make(chan struct{}, 1),
+		errChan:  make(chan error, 1),
+		wg:       &sync.WaitGroup{},
+	}
+}
+
+// Start starts the collector
+func (s StandaloneCollectorService) Start(ctx context.Context) error {
+	collectorStartedChan := make(chan struct{})
+	go func() {
+		// Start's context is only valid for the lifetime of "Start",
+		// but the collector expects a context that is valid for the lifetime of the service.
+		err := s.col.Run(context.Background())
+		if err != nil {
+			s.errChan <- err
+		}
+		close(collectorStartedChan)
+	}()
+
+	select {
+	case <-collectorStartedChan: // OK
+	case <-ctx.Done():
+		return fmt.Errorf("failed while waiting for service startup: %w", ctx.Err())
+	}
+
+	// monitor status for errors, so we don't zombie the service
+	s.wg.Add(1)
+	go s.monitorStatus(s.wg)
+	return nil
+}
+
+// monitorStatus monitors the collector's status for errors, and reports them
+// to the error channel to trigger a shutdown.
+func (s StandaloneCollectorService) monitorStatus(wg *sync.WaitGroup) {
+	defer wg.Done()
+	statusChan := s.col.Status()
+	for {
+		select {
+		case status := <-statusChan:
+			if status.Err != nil {
+				s.errChan <- status.Err
+			} else if !status.Running {
+				// If we aren't running, bail out. Otherwise the collector is effectively a "zombie" process.
+				s.errChan <- errors.New("collector unexpectedly stopped running")
+			}
+		case <-s.doneChan:
+			return
+		}
+	}
+}
```

Comment:
This might not be needed anymore. I fixed the collector in OTel a while back so that Run will return an error if anything is wrong during startup. It'll also set the status to `Closed` if startup fails instead of leaving it as `Starting` like previously.

Reply:
So ideally you just kick off `col.Run` in a goroutine and if it returns an error you notify the caller.

Reply:
Hmm, you might be right. I think the reason why I have this is context cancellation, actually. If the context passed into s.col.Run is cancelled unexpectedly, no error is returned, and the process becomes a zombie. 

I put this as a failsafe if there were any similar scenarios where the underlying collector might stop but not report an error.

Reply:
Might be something to talk through. This logic was originally a big protection as the collector would die and still report it's attempting to start.